### ISSUE TITLE
Feedback: 'Licensing'

### ISSUE THREAD

**Issue by mauroservienti** [Particular/docs.particular.net#3654]
Feedback for 'Licensing' https://docs.particular.net/nservicebus/licensing/

The page mentions the locations where the licence file can be saved:

```
{SpecialFolder.LocalApplicationData}\ParticularSoftware\license.xml
{SpecialFolder.CommonApplicationData}\ParticularSoftware\license.xml
```

The above paths are not copy/pastable and in order to understand what `LocalApplicationData` or `CommonApplicationData` is one needs to google for them or to write some code.

It's probably better to express these paths using something like:

```
%LOCALAPPDATA%\ParticularSoftware...
```

(and the corresponding path on Linux if needed)
so that the value can be copy/pasted to point directly to the directory

Location in GitHub: https://github.com/Particular/docs.particular.net/blob/master/nservicebus/licensing/index.md

ping @Particular/nservicebus-maintainers

**Comment by timbussmann** [comment]
I brought this up once as well and discussed this with @bording , the problem is that those paths are 
a.) different defaults on different OSes and 
b.) can easily be configured differently by the system administrator to point to other locations

but I agree that this doco doesn't help much to locate the used paths. Maybe we could document some of the standard defaults, or maybe log the searched path when NServiceBus doesn't find a license, so you just look at the log output and it tells you where to put the license (if we do not already do that...)?

**Comment by dbelcham** [comment]
I worked with this stuff last week and at first had thoughts about the docs not being clear. I do think, however, that adding more information about the full path is us documenting the .NET framework for MS. Instead maybe we should link off to a MSDN page that details what those special folders are.

**Comment by mauroservienti** [comment]
Agreed, the problem I faced while looking for those path this morning is that [MSDN documentation for .NET](https://msdn.microsoft.com/en-us/library/system.environment.specialfolder(v=vs.110).aspx) doesn't provide any useful information, but points to the [Windows SDK documentation](https://msdn.microsoft.com/library/dd378457.aspx) that provides what the user needs, but makes it hard to correlate from `SpecialFolder` enum values to Windows `KNOWNFOLDERID`.

**Comment by dbelcham** [comment]
I think I ended up on this page (https://johnkoerner.com/csharp/special-folder-values-on-windows-versus-mac/) which shows Windows and Mac (but not Linux) values for the `SpecialFolder` enum. I don't think this is a page we'd want to link to in our doco though.

Maybe we do need to document something like the above link has (for Windows, MacOS and Linux though) for a set of the folders that we care about?

**Comment by bording** [comment]
> maybe log the searched path when NServiceBus doesn't find a license

@timbussmann we already do this at the DEBUG log level.

@mauroservienti What everyone has said here so far is the main reason the page is the way it is currently.

Even on Windows, the actual paths represented by the SpecialFolder enum aren't guaranteed to be known, when Group Policy and folder redirection are involved, they can end up anywhere.

On mac and linux, .NET Core follows the [XDG spec](https://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html) which also allows for the locations to overridden.

The question is, how much of that do we actually want to document?

**Comment by timbussmann** [comment]
I feel like it wouldn't hurt to state the most common default location with a note describing that the locations might differ and how to see the searched locations via logging.
Also maybe we should not use Debug as a log level for that as we log info by default?

**Comment by bording** [comment]
> Also maybe we should not use Debug as a log level for that as we log info by default?

That is what I initially had, but we ended up changing it to Debug to cut down on log noise.

**Comment by mauroservienti** [comment]
Aren't those the type of things I as a user beginning with NServiceBus would like to see in the logs?

@sergioc @DavidBoike @dvdstelt @HEskandari this sounds like a good fit for us

**Comment by timbussmann** [comment]
@mauroservienti keep in mind that we automatically generate and use a trial license when beginning. I think reviewing the debug vs. info log level might still be a worthwhile discussions tbh.

**Comment by mauroservienti** [comment]
I know but that IMO complicates things 😄 

* You have a license file installed in the registry for example, because you’re already using V6
* Create a new project with V7/NET Core
* Press F5
* License cannot be found
* Trial is generated
* a browser window is opened talking about trial licenses

You have no easy way to understand what to do to fix the issue, and once realized that the license file cannot be read anymore from the registry you’re left alone finding where to put te license file.

**Comment by timbussmann** [comment]
@mauroservienti when running on the full framework, v7 will continue to detect licenses located in the registry (that will likely go away once we stop multi-targeting though).

**Comment by mauroservienti** [comment]
Yup, on full fx it gos on working. On NET Core it doesn’t, the latter was my case.

**Comment by timbussmann** [comment]
is that something to be solely addressed via docs? Maybe that worth raising as a pdev issue or to discuss with the @Particular/dotnetcore-launch-squad ?

**Comment by mauroservienti** [comment]
Good point @timbussmann

**Comment by DavidBoike** [comment]
I raised the PR https://github.com/Particular/docs.particular.net/pull/3664 today for completely unrelated reasons, which at least partially addresses this issue.

**Comment by weralabaj** [comment]
ping, what's the status of this issue? did @DavidBoike 's PR addressed all problems or is there anything left?

**Comment by mauroservienti** [comment]
To me it’s good enough, the PR adds license locations using Windows env variables, making it easy to copy paste those into a console window to get the actual path on the user machine.

I’ll close this as fixed by #3664. We can always raise another improvement.
